bitcoin/src/address/error.rs

```diff
@@ -7,18 +7,17 @@ use internals::write_err;
 use crate::address::{Address, NetworkUnchecked};
 use crate::blockdata::script::{witness_program, witness_version};
 use crate::prelude::*;
-use crate::Network;
 
 /// Address's network differs from required one.
 #[derive(Debug, Clone, PartialEq, Eq)]
-pub struct NetworkValidationError {
+pub struct NetworkValidationError<N> {
     /// Network that was required.
-    pub(crate) required: Network,
+    pub(crate) required: N,
     /// The address itself.
     pub(crate) address: Address<NetworkUnchecked>,
 }
```

Comment:
In 454296364d33c45d998b7678da221b6a41cc00d0:

This looks much worse. We now have a generic error type, which basically cannot be composed without the generic propagating through the entire tree of errors, and it's not clear what we actually need from the generic (apparently: it can be converted both to a main/test boolean and also to a HRP).

We should probably just stick a concrete `Params` in here if we mean the generic to be "some arbitrary network type".

Reply:
Honestly I think for `require_network` we just really need to accept `Network`. I don't see how we can reasonably write it such that we can use `Params` or anything else.

Reply:
You may be  correct. Though I could be convinced otherwise ... in principle, we could take an `AsRef<Params>` and then check the string address prefix against the `hrp` field or whatever.

In any case, people who want to make their own enums can work around this by writing their own checking function that calls `assume_checked` internally so I'm not too put out if it turns out we need `Network`.

Reply:
> we could take an `AsRef<Params>` and then check the string address prefix against the `hrp` field or whatever.

I was thinking about this since I believe we will have to do it anyway to deal with `non_exhaustive`.  But I wasn't sure how to then put the network into the error without making things generic. But now I've realized we can put `&'static str` there taking it from `network_name` field of `Params`.

> In any case, people who want to make their own enums can work around this by writing their own checking function that calls `assume_checked` internally so I'm not too put out if it turns out we need `Network`.

Yeah, and for enums that are subset of `Network` they can just convert (we can take `impl Into<Network>`). For non-subsets they already have other issues like dealing with HRP so maybe it's in unsupported category.